RxTodo/Sources/ViewControllers/TaskEditViewModel.swift

```diff
@@ -10,159 +10,144 @@ import RxCocoa
-          return .update(newTask)
-        }
-      }
-      .takeUntil(self.viewDidDeallocate)
-      .bindTo(provider.taskService.event)
-
-    //
-    // Dismiss
-    //
-    let cancelButtonDidTapWithoutChanges = self.cancelButtonItemDidTap
-      .withLatestFrom(self.titleInputText)
-      .filter { !cls.isTitleChanged(mode: mode, title: $0) }
-      .map { _ in Void() }
-
-    let cancelAlertDidSelectLeaveAction = self.cancelAlertDidSelectAction
-      .filter { $0 == .leave }
-      .map { _ in Void() }
-
-    let doneButtonDidTapWhenEnabled = self.doneButtonItemDidTap
-      .withLatestFrom(self.doneButtonEnabled)
-      .filter { isEnabled in isEnabled }
-      .map { _ in Void() }
-
-    self.dismissViewController = Observable
-      .of(cancelButtonDidTapWithoutChanges, cancelAlertDidSelectLeaveAction, doneButtonDidTapWhenEnabled)
-      .merge()
-      .observeOn(MainScheduler.instance)
-      .subscribeOn(ConcurrentMainScheduler.instance)
-  }
-
-
-  // MARK: - Functions
-
-  class func navigationBarTitle(mode: TaskEditViewMode) -> Driver<String?> {
+struct TaskEditViewModelOutputs {
+    var navigationBarTitle: Driver<String?>
+    var doneButtonEnabled: Driver<Bool>
+    var titleInputText: Driver<String?>
+    var presentCancelAlert: Observable<[TaskEditViewCancelAlertAction]>
+    var dismissViewController: Observable<Void>
+}
+
+typealias TaskEditViewModel = (TaskEditViewModelInputs) -> TaskEditViewModelOutputs
+
+func createTaskEditViewModel(provider: ServiceProviderType, mode: TaskEditViewMode) -> TaskEditViewModel {
+    return { input in
+      let cancelAlertDidSelectAction = input.cancelAlertDidSelectAction.asInput()
+      let cancelButtonItemDidTap = input.cancelButtonItemDidTap.asInput()
+      let doneButtonItemDidTap = input.doneButtonItemDidTap.asInput()
+      let titleInputDidChangeText = input.titleInputDidChangeText.asInput()
+
+        //
+        // Title Input Text
+        //
+        let titleInputText = createTitleInputText(
+            mode: mode,
+            titleInputDidChangeText: titleInputDidChangeText.asObservable()
+        )
+
+        //
+        // Navigation Item
+        //
+        let navigationBarTitle = createNavigationBarTitle(mode: mode)
+
+        let doneButtonEnabled = titleInputText
+            .map { text in text?.isEmpty == false }
+            .startWith(false)
+            .distinctUntilChanged()
+            .asDriver(onErrorJustReturn: false)
+
+        //
+        // Confirm Cancel
+        //
+        let presentCancelAlert: Observable<[TaskEditViewCancelAlertAction]> = cancelButtonItemDidTap
+            .withLatestFrom(titleInputText)
+            .filter { isTitleChanged(mode: mode, title: $0) }
+            .map { _ in [.leave, .stay] }
+            .observeOn(MainScheduler.instance)
+            .subscribeOn(ConcurrentMainScheduler.instance)
+
+        //
+        // Done
+        //
+        let createTaskEvent = doneButtonItemDidTap
```

Comment:
Yes you're right. I think one way to handle this is to recreate the view model every time the view is loaded; then we could remove the calls to asInput(), which is causing this problem by removing the propagation of Completion events from the UI controls.

Or, it would be possible to watch for deallocation events, though that still seems a bit messy to me. I'll play with it some more.